pyiceberg/table/update/validate.py

```diff
@@ -0,0 +1,72 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Optional
+
+from pyiceberg.manifest import ManifestContent, ManifestFile
+from pyiceberg.table import Table
+from pyiceberg.table.snapshots import Operation, Snapshot, ancestors_between
+
+
+class ValidationException(Exception):
+    """Raised when validation fails."""
+
+
+def validation_history(
+    table: Table,
+    from_snapshot: Snapshot,
+    to_snapshot: Optional[Snapshot],
+    matching_operations: set[Operation],
+    manifest_content_filter: ManifestContent,
+) -> tuple[list[ManifestFile], set[Snapshot]]:
+    """Return newly added manifests and snapshot IDs between the starting snapshot and parent snapshot.
+
+    Args:
+        table: Table to get the history from
+        from_snapshot: Parent snapshot to get the history from
+        to_snapshot: Starting snapshot
+        matching_operations: Operations to match on
+        manifest_content_filter: Manifest content type to filter
+
+    Raises:
+        ValidationException: If no matching snapshot is found or only one snapshot is found
+
+    Returns:
+        List of manifest files and set of snapshots matching conditions
+    """
+    manifests_files: list[ManifestFile] = []
+    snapshots: set[Snapshot] = set()
+
+    last_snapshot = None
+    for snapshot in ancestors_between(from_snapshot, to_snapshot, table.metadata):
+        last_snapshot = snapshot
+        summary = snapshot.summary
+        if summary is None or summary.operation not in matching_operations:
```

Comment:
This looks dangerous to me, I don't think we want to skip over this if the summary is `None`

Reply:
I think the right thing here is to assume that it is an `Overwrite` operation

Reply:
What is a case where summary would ever be `None`? I can't seem to find any checks against that in the codebase except for https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/snapshots.py#L367 which seems to be checking against an empty dict (well mapping) and should also be checking the `or is None` case imo

Reply:
It now defaults to `Operation.OVERWRITE` if summary is determined to be `None`, but still wondering about above

Reply:
`summary` can be `None` if the Table format version is `v1` according to the Iceberg Spec: https://iceberg.apache.org/spec/#snapshots

Instead of setting it as `Operation.OVERWRITE` I think an alternative approach is to just throw an exception if `summary` field is None with a helpful error message that explains that `validation_history` cannot be generated in the absence of summary fields. Wdyt @Fokko @jayceslesar ?

Reply:
@sungwy That would also work for me. We can always see if there are ways to make it work for V1 tables later, however, I'm not sure how big the problem is because most implementations also produce an operation for V1 tables. So, I think we're good with the approach that @sungwy suggests 👍

Reply:
Sounds good to me, implemented and added to the existing test to assure that we raise an error when that happens. Planning on cleaning up tests in the next MR when I will likely need to make some fixtures & parameterize